### ISSUE TITLE
Telegraf is consuming a lot of CPU on server with 120K established connections

### ISSUE THREAD

**Issue by jimis** [influxdata/telegraf#6559]
telegraf-1.12.3 with data collection interval 10s, sending data to influxDB
OS: Linux x86_64

* Telegraf was consuming constantly around 70% CPU (on average, as measured by telegraf itself with the procstat plugin). 
* Server has around 120K established incoming connections because of other services running
* The CPU usage went back low when the connections were gone
* I am suspecting the culprit is "**inputs.netstat**" plugin which is enabled with default options

**Comment by danielnelson** [comment]
@jimis Would you be able to help us collect a few profiles?  Our [documentation for profiling][1] needs a bit of work first though, we will get back to you with an update that has more precise instructions.

[1]: https://github.com/influxdata/telegraf/blob/master/docs/PROFILING.md

**Comment by jimis** [comment]
Yes I can do some profiling. I already tried to grab some gdb backtraces to see where the threads are kept busy, but all I got was question marks instead of function names. Are you binaries shipping with debug info?

FYI, on this server it is InfluxDB that has thousands of connections in ESTABLISHED state. There is an issue there, because I see a slow but constant upwards trend in the number of connections (as logged by telegraf), and the number gets reset only when I restart influxd. I hope I'll fix this issue soon, but until then I can definitely profile telegraf.

**Comment by jimis** [comment]
I don't think profiling is necessary, I have verified that the slowness is in the **inputs.netstat** plugin. 

What I did to verify this:
* disabled netstat plugin and restarted telegraf
* CPU utilisation went very low
* re-enable netstat plugin and restart telegraf
* CPU utilisation goes to 70% on average

I believe it should be easy for you to replicate the scenario, by providing a fake `/proc/net/tcp6` file with 120K lines. I can provide you with one such file if needed.

**Comment by jimis** [comment]
@glinton I see you added the upstream label, but I have to disagree, after reading a bunch of comments in issues in gopsutil. gopsutil is having the job to parse and return information from every single process in the system. While telegraf can do much better by opening only one generic file.

@danielnelson I see you removed the `ready` label, what more information would you need from me? I don't think profiling will be useful, the culprit was easy to identify.

**Comment by danielnelson** [comment]
I'll answer over on the gopsutil issue too, it may make sense for us to have a custom linux implementation because gopsutil is reporting the pid, and we don't need it in this plugin.  I don't think the pid can be obtained from `/proc/net/tcp` and it appears from strace that when given the `-p` flag netstat follows a similar strategy to gopsutil.

**Comment by danielnelson** [comment]
If you can work on an implementation that uses the `/proc/net/tcp*` files, we can either add it to Telegraf or into gopsutil depending on what is decided.

**Comment by JoveYu** [comment]
I can get tcp netstat from `/proc/net/tcp*`, but I don't know what the metrics `udp_socket` means, and where is it from?

**Comment by jimis** [comment]
> I can get tcp netstat from `/proc/net/tcp*`, but I don't know what the metrics `udp_socket` means, and where is it from?

@JoveYu I would /guess/ it is the number of udp sockets the system keeps track of. Probably the number of lines of some of the files  `/proc/net/udp*` ?

**Comment by JoveYu** [comment]
i hava rewrite netstat plugin in linux, it can help fix this issue

**Comment by Kedesai** [comment]
@danielnelson thanks for pointing me to this open ticket.

**Comment by Kedesai** [comment]
What version of telegraf client has the fix @JoveYu and @danielnelson

**Comment by danielnelson** [comment]
Still working on the fix in #6772.

**Comment by Kedesai** [comment]
on the 15.1 telegraf version the netstat is consuming atleast 10 % cpu when netstat is enabled. On that server I have shut off the procstat and logparser plugin. I think it is extreme that when netstat is enabled it consumes so much cpu

When netstat enabled:
17959 telegraf  20   0 1349384  51156  19084 S  **11.9**  0.3   0:04.01 telegraf

When netstat disabled:
22460 telegraf  20   0 1275396  37704  18532 S   **6.3**  0.2   0:00.24 telegraf

This is extreme.

**Comment by Kedesai** [comment]
@danielnelson

**Comment by danielmotaleite** [comment]
@JoveYu and @danielnelson , the https://github.com/influxdata/telegraf/pull/6772 was closed without merging. Any info about any backstage changes to get this improved? we need the netstat to track the high number of connections, but as we have many, telegraf consumes way too much resources to have it enabled. Basically, when this plugin would be useful is exactly where we can't use it.

If there is any plan, how about merging the above patch as a new linux_netstat plugin ?

**Comment by Kedesai** [comment]
Great!! and someone commented that they did not see this issue and they run 1k servers.  I wonder why the ticket closes without the customer verifying which I cannot understand. Poor practice.
